### PR TITLE
Fix orgmode link resolving

### LINKED ISSUE #28974: README.org casuing 500 internal server error
### Description

`[[file:]]` Org links are causing 500 internal server errors (similar to #17139):
```
An error occurred:

Render, orgmode.Render failed: could not write output: interface conversion: org.Node is org.Text, not org.RegularLink
```

The culprit lies in Org `file:` links, e.g., [[GPLv3 Badge][file:https://img.shields.io/badge/license-GPL_3-green.svg]]. This should show a GPLv3 badge, but instead triggers the error. Although omitting the `file:` still achieves the intended result without an error, it should be noted that this works just fine on GitHub and in the previous version of Gitea, i.e., v1.21.3.

P.S. Gitea demo site reproduction was not possible as I could not even log into the demo site.

### Gitea Version

1.21.4

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

https://gist.github.com/arifer612/2baa4b54e00392e0d7f0238a2344450f

### Screenshots

Screenshot from 2024-01-29 21-26-08

### Git Version

2.40.1

### Operating System

unRAID

### How are you running Gitea?

Run as a Docker container.

### Database

MySQL/MariaDB

### PR DESCRIPTION
**Pull request by wxiaoguang**
Fix #28974

Add some new tests and fix some legacy unclear tests.

### CODE REVIEW

File: modules/markup/orgmode/orgmode_test.go
Comment (arifer612) [inline modules/markup/orgmode/orgmode_test.go:1]: This test will fail to produce the expected result when I want to display a badge that gets directed to a webpage when clicked on. For example, `[[https://develop.spacemacs.org][file:https://cdn.rawgit.com/syl20bnr/spacemacs/442d025779da2f62fc86c2082703697714db6514/assets/spacemacs-badge.svg]]` should be displayed the badge at `https://cdn.rawgit.com/syl20bnr/spacemacs/442d025779da2f62fc86c2082703697714db6514/assets/spacemacs-badge.svg`, and direct the user to `https://develop.spacemacs.org` when clicked.

Comment (wxiaoguang) [reply modules/markup/orgmode/orgmode_test.go:1]: Nope, it works as expected. The test case already covers it:

<details>

image

</details>

Comment (wxiaoguang) [reply modules/markup/orgmode/orgmode_test.go:1]: I added a new case to address your concern: dc92cb17fc0db1288da4c59c2fa869f30547235b 

image

**Review by lunny — APPROVED** [review]

**Comment by KN4CK3R** [comment]
The markdown renderer had some problems with relative and absolute links (#15075, #28915) (fixed by #28932). The org mode renderer has the same problem. Should we fix that too? I think we don't have a bug report at the moment because only a few people use org mode.

```
- [[docs/running.md][relative]]
- [[/docs/running.md][absolute]]
- [[file:docs/running.md][relative file]]
- [[file:/docs/running.md][absolute file]]
```

**Comment by wxiaoguang** [comment]
I think this PR's main purpose is to "fix" the 500 error (the `[[link][description]]` syntax).

Feel free to edit it and add more fixes (eg: "relative and absolute links"), maybe the fixes could also be done in a separate PR.

Either is fine to me.

**Comment by KN4CK3R** [comment]
If branch infos are present, the wrong url is generated:
```
Links: markup.Links{
	Base: "/relative-path",
	BranchPath: "branch/main",
},

test("[[WikiPage][The WikiPage Name]]", `<p><a href="/relative-path/src/branch/main/WikiPage">The WikiPage Name</a></p>`)
// <p><a href="/relative-path/media/branch/main/WikiPage">The WikiPage Name</a></p>
```
The generated path uses the `media` path but the link is no media.

**Comment by wxiaoguang** [comment]
> If branch infos are present, the wrong url is generated:
> The generated path uses the media path but the link is no media.

Hmm, I see. I think I understand the old bug now. By old code, in many cases the "relative link" is still incorrect, because orgmode reports `link Kind == regular` for `[[ImageLink.svg][The Image Desc]]`. 

So, either I revert the code to that old (bug buggy) behavior, or add more code to "fix". What do you think ?

**Comment by wxiaoguang** [comment]
Made some "fixes" in 6cd906bdf0ad83ada245c21dbd3447f663bf3541, are they right now?

```
	test("[[WikiPage][The WikiPage Desc]]",
		`<p><a href="/relative-path/WikiPage">The WikiPage Desc</a></p>`)
	test("[[ImageLink.svg][The Image Desc]]",
		`<p><a href="/relative-path/media/branch/main/ImageLink.svg">The Image Desc</a></p>`)
```

**Comment by KN4CK3R** [comment]
The following may be a different PR.

Do we want the orgmode renderer behave like the markdown renderer? I would say yes but then we may have to put more work in this. I created multiple test cases extracted from the markdown renderer:

```go
	input := `/just/a/path.bin
https://example.com/file.bin
[[file:file.bin][local link]]
[[https://example.com][remote link]]
[[file:image.jpg][local image]]
[[file:path/file.jpg][local image]]
[[file:/path/file.jpg][local image]]
[[file:https://example.com/image.jpg][remote image]]`

	cases := []struct {
		Links    markup.Links
		Expected string
	}{
		{ // 0
			Links:  markup.Links{},
			Expected: `<p>/just/a/path.bin<br/>
<a href="https://example.com/file.bin" rel="nofollow">https://example.com/file.bin</a><br/>
<a href="/file.bin" rel="nofollow">local link</a><br/>
<a href="https://example.com" rel="nofollow">remote link</a><br/>
<a href="/image.jpg" target="_blank" rel="nofollow noopener"><img src="/image.jpg" alt="local image"/></a><br/>
<a href="/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/path/file.jpg" alt="local image"/></a><br/>
<a href="/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/path/file.jpg" alt="local image"/></a><br/>
<a href="https://example.com/image.jpg" target="_blank" rel="nofollow noopener"><img src="https://example.com/image.jpg" alt="remote image"/></a></p>`,
		},
		{ // 1
			Links:  markup.Links{
				Base:"https://gitea.io/",
			},
			Expected: `<p>/just/a/path.bin<br/>
<a href="https://example.com/file.bin" rel="nofollow">https://example.com/file.bin</a><br/>
<a href="https://gitea.io/file.bin" rel="nofollow">local link</a><br/>
<a href="https://example.com" rel="nofollow">remote link</a><br/>
<a href="https://gitea.io/image.jpg" target="_blank" rel="nofollow noopener"><img src="https://gitea.io/image.jpg" alt="local image"/></a><br/>
<a href="https://gitea.io/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="https://gitea.io/path/file.jpg" alt="local image"/></a><br/>
<a href="https://gitea.io/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="https://gitea.io/path/file.jpg" alt="local image"/></a><br/>
<a href="https://example.com/image.jpg" target="_blank" rel="nofollow noopener"><img src="https://example.com/image.jpg" alt="remote image"/></a></p>`,
		},
		{ // 2
			Links:  markup.Links{
				Base: "/relative/path",
			},
			Expected: `<p>/just/a/path.bin<br/>
<a href="https://example.com/file.bin" rel="nofollow">https://example.com/file.bin</a><br/>
<a href="/relative/path/file.bin" rel="nofollow">local link</a><br/>
<a href="https://example.com" rel="nofollow">remote link</a><br/>
<a href="/relative/path/image.jpg" target="_blank" rel="nofollow noopener"><img src="/relative/path/image.jpg" alt="local image"/></a><br/>
<a href="/relative/path/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/relative/path/path/file.jpg" alt="local image"/></a><br/>
<a href="/relative/path/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/relative/path/path/file.jpg" alt="local image"/></a><br/>
<a href="https://example.com/image.jpg" target="_blank" rel="nofollow noopener"><img src="https://example.com/image.jpg" alt="remote image"/></a></p>`,
		},
		{ // 3
			Links:  markup.Links{
				Base:       "/user/repo",
				BranchPath: "branch/main",
			},
			Expected: `<p>/just/a/path.bin<br/>
<a href="https://example.com/file.bin" rel="nofollow">https://example.com/file.bin</a><br/>
<a href="/user/repo/src/branch/main/file.bin" rel="nofollow">local link</a><br/>
<a href="https://example.com" rel="nofollow">remote link</a><br/>
<a href="/user/repo/media/branch/main/image.jpg" target="_blank" rel="nofollow noopener"><img src="/user/repo/media/branch/main/image.jpg" alt="local image"/></a><br/>
<a href="/user/repo/media/branch/main/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/user/repo/media/branch/main/path/file.jpg" alt="local image"/></a><br/>
<a href="/user/repo/media/branch/main/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/user/repo/media/branch/main/path/file.jpg" alt="local image"/></a><br/>
<a href="https://example.com/image.jpg" target="_blank" rel="nofollow noopener"><img src="https://example.com/image.jpg" alt="remote image"/></a></p>`,
		},
		{ // 4
			Links:  markup.Links{
				Base:     "/relative/path",
				TreePath: "sub/folder",
			},
			Expected: `<p>/just/a/path.bin<br/>
<a href="https://example.com/file.bin" rel="nofollow">https://example.com/file.bin</a><br/>
<a href="/relative/path/file.bin" rel="nofollow">local link</a><br/>
<a href="https://example.com" rel="nofollow">remote link</a><br/>
<a href="/relative/path/image.jpg" target="_blank" rel="nofollow noopener"><img src="/relative/path/image.jpg" alt="local image"/></a><br/>
<a href="/relative/path/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/relative/path/path/file.jpg" alt="local image"/></a><br/>
<a href="/relative/path/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/relative/path/path/file.jpg" alt="local image"/></a><br/>
<a href="https://example.com/image.jpg" target="_blank" rel="nofollow noopener"><img src="https://example.com/image.jpg" alt="remote image"/></a></p>`,
		},
		{ // 5
			Links:  markup.Links{
				Base:       "/user/repo",
				BranchPath: "branch/main",
				TreePath:   "sub/folder",
			},
			Expected: `<p>/just/a/path.bin<br/>
<a href="https://example.com/file.bin" rel="nofollow">https://example.com/file.bin</a><br/>
<a href="/user/repo/src/branch/main/sub/folder/file.bin" rel="nofollow">local link</a><br/>
<a href="https://example.com" rel="nofollow">remote link</a><br/>
<a href="/user/repo/media/branch/main/sub/folder/image.jpg" target="_blank" rel="nofollow noopener"><img src="/user/repo/media/branch/main/sub/folder/image.jpg" alt="local image"/></a><br/>
<a href="/user/repo/media/branch/main/sub/folder/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/user/repo/media/branch/main/sub/folder/path/file.jpg" alt="local image"/></a><br/>
<a href="/user/repo/media/branch/main/sub/folder/path/file.jpg" target="_blank" rel="nofollow noopener"><img src="/user/repo/media/branch/main/sub/folder/path/file.jpg" alt="local image"/></a><br/>
<a href="https://example.com/image.jpg" target="_blank" rel="nofollow noopener"><img src="https://example.com/image.jpg" alt="remote image"/></a></p>`,
		},
	}

	for i, c := range cases {
		result, err := RenderString(&markup.RenderContext{Ctx: context.Background(), Links: c.Links}, input)
		assert.NoError(t, err, "Unexpected error in testcase: %v", i)
		assert.Equal(t, c.Expected, result, "Unexpected result in testcase %v", i)
	}
```

The corresponding markdown input looks like this:
```go
input := `/just/a/path.bin
https://example.com/file.bin
[local link](file.bin)
[remote link](https://example.com)
![local image](image.jpg)
![local image](path/file.jpg)
![local image](/path/file.jpg)
![remote image](https://example.com/image.jpg)`
```

Things I noticed:
- `\n` gets rendered as `<br/>\n` in markdown and just `\n` in orgmode
- links in markdown get decorated with `target` and `rel` attributes.

**Comment by wxiaoguang** [comment]
The test cases are really helpful, thank you very much.

> The following may be a different PR.

So if I understand correctly, this PR doesn't need to introduce more changes at the moment?

**Comment by KN4CK3R** [comment]
Yes, want to let you decide if you want to make changes here or we use another PR to adjust the code. It's different to markdown for a long time, no need to hurry.

**Review by KN4CK3R — APPROVED** [review]

**Pull request merged**